rx-preferences/src/main/java/com/f2prateek/rx/preferences2/RealPreference.java

```diff
@@ -65,22 +67,21 @@
     return adapter.get(key, preferences);
   }
 
-  @Override public synchronized void set(@Nullable T value) {
+  @Override public synchronized void set(@NonNull T value) {
+    checkNotNull(value, "value == null");
     SharedPreferences.Editor editor = preferences.edit();
-    if (value == null) {
-      editor.remove(key);
-    } else {
-      adapter.set(key, value, editor);
-    }
+    adapter.set(key, value, editor);
     editor.apply();
   }
 
-  @Override public boolean isSet() {
+  @Override public synchronized boolean isSet() {
```

Comment:
https://github.com/f2prateek/rx-preferences/issues/77#issuecomment-288469121

Here's a sequence that exhibits what those guards prevent:

a) User calls #get(key). RealPreference verifies existence of `key`.
b) User deletes `key` before the adapter is called.
c) Preference.Adapter#get is called. It now retrieves the value for `key` from the underlying preferences.

Given the check in a), the adapter in step c) assumes that a value for `key` is present. But b) deleted the value. This means the adapter would return `null`.

By adding the guard, we guarantee that b) waits until c) is done before executing.